socs/db/suprsync.py

```diff
@@ -0,0 +1,394 @@
+            Create table if it hasn't been generated yet.
+        echo : bool
+            If true, writes sql statements to stdout
+    """
+    def __init__(self, db_path, create_all=True, echo=False):
+        db_path = os.path.abspath(db_path)
+        if not os.path.exists(os.path.dirname(db_path)):
+            os.makedirs(os.path.dirname(db_path))
+
+        self._engine = create_engine(f'sqlite:///{db_path}', echo=echo)
+        self.Session = sessionmaker(bind=self._engine)
+
+        if create_all:
+            Base.metadata.create_all(self._engine)
+
+    def add_file(self, local_path, remote_path, archive_name,
+                 local_md5sum=None, timestamp=None, session=None,
+                 deletable=True):
+        """
+        Adds file to the SupRsyncFiles table.
+
+        Args
+        ----
+            local_path : String
+                Absolute path of the local file to be copied
+            remote_path : String
+                Path of the file on the remote server relative to the base-dir.
+                specified in the SupRsync agent config.
+            archive_name : String
+                Name of the archive, i.e. `timestreams` or `smurf`. Each archive
+                is managed by its own SupRsync instance, so they can be copied to
+                different base-dirs or hosts.
+            local_md5sum : String, optional
+                locally calculated checksum. If not specified, will calculate
+                md5sum automatically.
+            session : sqlalchemy session
+                Session to use to add the SupRsyncFile. If None, will create
+                a new session and commit afterwards.
+            deletable : bool
+                If true, can be deleted by suprsync agent
+        """
+        file = create_file(local_path, remote_path, archive_name,
+                           local_md5sum=local_md5sum, timestamp=timestamp,
+                           deletable=deletable)
+        if session is None:
+            with self.Session.begin() as session:
+                session.add(file)
+        else:
+            session.add(file)
+
+    def get_copyable_files(self, archive_name, session=None,
+                           max_copy_attempts=None, num_files=None):
+        """
+        Gets all SupRsyncFiles that are copyable, meaning they satisfy:
+         - local and remote md5sums do not match
+         - Failed copy attempts is below the max number of attempts
+
+        Args
+        ----
+            archive_name (string):
+                Name of archive to get files from
+            session (sqlalchemy session):
+                Session to use to get files. If none is specified, one will
+                be created. You need to specify this if you wish to change
+                file data and commit afterwards.
+            max_copy_attempts (int):
+                Max number of failed copy atempts
+            num_files (int):
+                Number of files to return
```

Comment:
Formatting here is slightly off. Arguments should be "name : type". See the docstring for `add_file()` above. The current syntax gets mangled when compiled by Sphinx.

Reply:
ah this always confuses me. I swear I was just doing something where this was the correct format for docstrings but maybe in a different section or something...

Reply:
This is *a* correct format. I think the issue might be the mix of formats. The rest of the file is in the Numpy style. This is the Google style.

I constantly reference this page (and the linked Google style at the top): https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy